crates/data_components/src/s3_vectors/query_provider.rs

```diff
@@ -148,36 +185,115 @@ impl TableProvider for S3VectorsQueryTable {
 
     async fn scan(
         &self,
-        _state: &dyn Session,
+        state: &dyn Session,
         projection: Option<&Vec<usize>>,
         filters: &[Expr],
         limit: Option<usize>,
     ) -> DataFusionResult<Arc<dyn ExecutionPlan>> {
-        #[allow(clippy::cast_possible_wrap)]
-        let limit: i64 = match limit {
-            Some(l) if (l as i64) > S3_VECTOR_MAX_TOPK => {
-                tracing::warn!(
-                    "S3VectorsQueryTable: limit {l} exceeds maximum of {S3_VECTOR_MAX_TOPK}, truncating."
-                );
-                S3_VECTOR_MAX_TOPK
-            }
-            None => S3_VECTOR_MAX_TOPK,
-            Some(l) => l as i64,
-        };
-
         let query_vector = self
             .compute_vector
             .compute_vector(self.query.as_str())
             .await
             .map_err(DataFusionError::External)?;
 
-        Ok(Arc::new(S3VectorsQueryExec::new(
-            self,
-            projection,
-            limit,
-            query_vector,
-            filters.to_vec(),
-        )) as Arc<dyn ExecutionPlan>)
+        if self.partition_by.is_empty() {
+            return Ok(self.no_partitioning_exec(projection, filters, limit, query_vector));
+        }
+
+        let (_, Some(bucket_name), Some(index_name)) = self.table.idx.index_identifier_variables()
+        else {
+            return exec_err!("No bucket name or index name for bucket query");
+        };
+
+        let list_indexes_output = self
+            .table
+            .client
+            .list_indexes(
+                ListIndexesInput::builder()
```

Comment:
similar to my previous comment, this should use prefix filtering

Reply:
this logic is almost exactly the same as the ListTable one except for the underlying struct it creates - maybe worth a generic function or a function with a callback that creates the `S3VectorsQueryTable`/`S3VectorsListTable`